pkg/policies/controlplane/components/gradient-controller.go

```diff
@@ -0,0 +1,142 @@
+package components
+
+import (
+	"fmt"
+	"math"
+
+	"go.uber.org/fx"
+
+	policylangv1 "github.com/fluxninja/aperture/v2/api/gen/proto/go/aperture/policy/language/v1"
+	"github.com/fluxninja/aperture/v2/pkg/config"
+	"github.com/fluxninja/aperture/v2/pkg/notifiers"
+	"github.com/fluxninja/aperture/v2/pkg/policies/controlplane/iface"
+	"github.com/fluxninja/aperture/v2/pkg/policies/controlplane/runtime"
+)
+
+// GradientController describes gradient values.
+type GradientController struct {
+	lastOutput        runtime.Reading
+	minGradient       float64
+	maxGradient       float64
+	slope             float64
+	componentID       string
+	dynamicConfigKey  string
+	defaultManualMode bool
+	manualMode        bool
+}
+
+// Make sure Gradient complies with Component interface.
+var _ runtime.Component = (*GradientController)(nil)
+
+// NewGradientControllerAndOptions creates a Gradient Controller Component and its fx options.
+func NewGradientControllerAndOptions(
+	gradientControllerProto *policylangv1.GradientController,
+	componentID runtime.ComponentID,
+	policyReadAPI iface.Policy,
+) (runtime.Component, fx.Option, error) {
+	gradientParameters := gradientControllerProto.GetParameters()
+	// Make sure max is greater than min
+	if gradientParameters.MaxGradient < gradientParameters.MinGradient {
+		return nil, nil, fmt.Errorf("max_gradient must be greater than min_gradient")
+	}
+
+	gradient := &GradientController{
+		slope:             gradientParameters.Slope,
+		minGradient:       gradientParameters.MinGradient,
+		maxGradient:       gradientParameters.MaxGradient,
+		componentID:       componentID.String(),
+		dynamicConfigKey:  gradientControllerProto.ManualModeConfigKey,
+		defaultManualMode: gradientControllerProto.ManualMode,
+		manualMode:        gradientControllerProto.ManualMode,
+	}
+
+	return gradient, fx.Options(), nil
+}
+
+// Name implements runtime.Component.
+func (g *GradientController) Name() string {
+	return "Gradient"
+}
+
+// Type implements runtime.Component.
+func (g *GradientController) Type() runtime.ComponentType {
+	return runtime.ComponentTypeSignalProcessor
+}
+
+// ShortDescription implements runtime.Component.
+func (g *GradientController) ShortDescription() string {
+	return "Gradient Controller"
+}
+
+// IsActuator implements runtime.Component.
+func (g *GradientController) IsActuator() bool {
+	return false
+}
+
+// Execute implements runtime.Component.Execute.
+func (g *GradientController) Execute(inPortReadings runtime.PortToReading, circuitAPI runtime.CircuitAPI) (outPortReadings runtime.PortToReading, err error) {
+	signal := inPortReadings.ReadSingleReadingPort("signal")
+	setpoint := inPortReadings.ReadSingleReadingPort("setpoint")
+	optimize := inPortReadings.ReadSingleReadingPort("optimize")
```

Comment:
Don't have `optimize` port on `GradientController`

Reply:
[63c267f](https://github.com/fluxninja/aperture/pull/2838/commits/63c267ff138b9e85f4f18813a91aa8bf78e58e8e) addresses this.